compute/cker/src/train/DepthwiseConv.test.cc

```diff
@@ -0,0 +1,410 @@
+/*
+ * Copyright (c) 2024 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cker/train/operation/DepthwiseConv.h>
+
+#include <gtest/gtest.h>
+#include <vector>
+
+namespace
+{
+
+template <typename T> class DepthwiseConvVerifier
+{
+public:
+  DepthwiseConvVerifier() : _dconv_kernel{new nnfw::cker::train::DepthwiseConv()}
+  {
+    _dconv_kernel = std::make_unique<nnfw::cker::train::DepthwiseConv>();
+  }
+
+  void prepare(const nnfw::cker::Shape &incoming_shape, const nnfw::cker::Shape &filter_shape)
+  {
+    const int kPacketSize = _dconv_kernel->kPacketSize<T>();
+    const int batch = incoming_shape.Dims(0);
+    const int out_depth = incoming_shape.Dims(3);
+    const int filter_rows = filter_shape.Dims(1);
+    const int filter_cols = filter_shape.Dims(2);
+    const int filter_spatial_size = filter_rows * filter_cols;
+    const int padded_filter_inner_dim_size =
+      ((out_depth + kPacketSize - 1) / kPacketSize) * kPacketSize;
+
+    _use_padded_filter = (out_depth % kPacketSize) == 0 ? false : true;
+    {
+      nnfw::cker::Shape padded_filter_shape(
+        {batch, filter_spatial_size, padded_filter_inner_dim_size});
+      _padded_filter.resize(padded_filter_shape.FlatSize());
+    }
+
+    {
+      const int thread_count = _dconv_kernel->getThreadCount() + 1;
```

Comment:
The meaning of `+1` is for main thread. The `numThreads()` in the eigen library only returns the thread count except for the main thread. This function can be executed in the main thread depending on the situation. So I added `1` for main thread.

Reply:
I don't fully understand. Out-of-curiosity, I searched tensorflow code (e.g. depthwise_conv_op.cc), it seems to use worker thread counts. I don't find `+`, but I found the following:
```cc
auto worker_threads = *(ctx->device()->tensorflow_cpu_worker_threads());
```
I guess, the `thread_count` should be the number of worker threads.
Is the main thread is part of worker threads?

Reply:
Is there any reason to add 1 for main thread in the thread count? As far as I know, main thread is usually not used in optimized kernel.

Reply:
> Is there any reason to add 1 for main thread in the thread count? As far as I know, main thread is usually not used in optimized kernel.

Could you give a link related to `main thread is usually not used in optimized kernel`? This is because when I tested this kernel using several cases, the main thread was also used. Especially if the batch size is 1, it usually uses the main thread. And in other cases, which thread is being used depends on the situation.

https://github.com/libigl/eigen/blob/1f05f51517ec4fd91eed711e0f89e97a7c028c0e/unsupported/Eigen/CXX11/src/Tensor/TensorDeviceThreadPool.h#L175-L179
```
    if (n <= 1 || numThreads() == 1 ||
        CostModel::numThreads(n, cost, static_cast<int>(numThreads())) == 1) {
      f(0, n);
      return;
    }
```
In this code, n is a batch_size. And the batch size is 1, it works on main thread.

```
(gdb) info threads
  Id   Target Id                                       Frame 
  1    Thread 0x7ffff7a48780 (LWP 1326007) "test_cker" nnfw::cker::depthwise_conv_op::LaunchDepthwiseConvBackpropInputOp<Eigen::ThreadPoolDevice, float>::operator()(int, int, int, int, int, int, int, int, int, int, int, int, int, float const*, float const*, float*, float*, bool, float*, float*)::{lambda(long, long)#1}::operator()(long, long) const (this=0x555555b0c560, start=0, limit=1) at /home/jyoung/git/npu/ONE/compute/cker/include/cker/eigen/depthwise_conv_op.h:592
  2    Thread 0x7ffff797e700 (LWP 1326011) "test_cker" futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x555555b0c9d8) at ../sysdeps/nptl/futex-internal.h:183
  3    Thread 0x7ffff717d700 (LWP 1326012) "test_cker" futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x555555b0ca58) at ../sysdeps/nptl/futex-internal.h:183
  4    Thread 0x7ffff697c700 (LWP 1326013) "test_cker" futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x555555b0cad8) at ../sysdeps/nptl/futex-internal.h:183
  5    Thread 0x7ffff617b700 (LWP 1326014) "test_cker" futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x555555b0cb58) at ../sysdeps/nptl/futex-internal.h:183
  6    Thread 0x7ffff597a700 (LWP 1326015) "test_cker" futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x555555b0cbd8) at ../sysdeps/nptl/futex-internal.h:183
  7    Thread 0x7ffff5179700 (LWP 1326016) "test_cker" futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x555555b0cc58) at ../sysdeps/nptl/futex-internal.h:183
  8    Thread 0x7ffff4978700 (LWP 1326017) "test_cker" futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x555555b0ccd8) at ../sysdeps/nptl/futex-internal.h:183
  9    Thread 0x7ffff4177700 (LWP 1326018) "test_cker" futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x555555b0cd58) at ../sysdeps/nptl/futex-internal.h:183
  10   Thread 0x7ffff3976700 (LWP 1326019) "test_cker" futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x555555b0cdd8) at ../sysdeps/nptl/futex-internal.h:183
  11   Thread 0x7ffff3175700 (LWP 1326020) "test_cker" futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x555555b0ce58) at ../sysdeps/nptl/futex-internal.h:183
  12   Thread 0x7ffff2974700 (LWP 1326021) "test_cker" futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x555555b0ced8) at ../sysdeps/nptl/futex-internal.h:183
  13   Thread 0x7ffff2173700 (LWP 1326022) "test_cker" nnfw::cker::depthwise_conv_op::LaunchDepthwiseConvBackpropInputOp<Eigen::ThreadPoolDevice, float>::operator()(int, int, int, int, int, int, int, int, int, int, int, int, int, float const*, float const*, float*, float*, bool, float*, float*)::{lambda(long, long)#1}::operator()(long, long) const (this=0x555555b0c560, start=8, limit=9) at /home/jyoung/git/npu/ONE/compute/cker/include/cker/eigen/depthwise_conv_op.h:592
  14   Thread 0x7ffff1972700 (LWP 1326023) "test_cker" nnfw::cker::depthwise_conv_op::LaunchDepthwiseConvBackpropInputOp<Eigen::ThreadPoolDevice, float>::operator()(int, int, int, int, int, int, int, int, int, int, int, int, int, float const*, float const*, float*, float*, bool, float*, float*)::{lambda(long, long)#1}::operator()(long, long) const (this=0x555555b0c560, start=9, limit=10) at /home/jyoung/git/npu/ONE/compute/cker/include/cker/eigen/depthwise_conv_op.h:592
  15   Thread 0x7ffff1171700 (LWP 1326024) "test_cker" nnfw::cker::depthwise_conv_op::LaunchDepthwiseConvBackpropInputOp<Eigen::ThreadPoolDevice, float>::operator()(int, int, int, int, int, int, int, int, int, int, int, int, int, float const*, float const*, float*, float*, bool, float*, float*)::{lambda(long, long)#1}::operator()(long, long) const (this=0x555555b0c560, start=6, limit=7) at /home/jyoung/git/npu/ONE/compute/cker/include/cker/eigen/depthwise_conv_op.h:592
  16   Thread 0x7ffff0970700 (LWP 1326025) "test_cker" nnfw::cker::depthwise_conv_op::LaunchDepthwiseConvBackpropInputOp<Eigen::ThreadPoolDevice, float>::operator()(int, int, int, int, int, int, int, int, int, int, int, int, int, float const*, float const*, float*, float*, bool, float*, float*)::{lambda(long, long)#1}::operator()(long, long) const (this=0x555555b0c560, start=4, limit=5) at /home/jyoung/git/npu/ONE/compute/cker/include/cker/eigen/depthwise_conv_op.h:592
  17   Thread 0x7ffff016f700 (LWP 1326026) "test_cker" nnfw::cker::depthwise_conv_op::LaunchDepthwiseConvBackpropInputOp<Eigen::ThreadPoolDevice, float>::operator()(int, int, int, int, int, int, int, int, int, int, int, int, int, float const*, float const*, float*, float*, bool, float*, float*)::{lambda(long, long)#1}::operator()(long, long) const (this=0x555555b0c560, start=3, limit=4) at /home/jyoung/git/npu/ONE/compute/cker/include/cker/eigen/depthwise_conv_op.h:592
  18   Thread 0x7fffef96e700 (LWP 1326027) "test_cker" nnfw::cker::depthwise_conv_op::LaunchDepthwiseConvBackpropInputOp<Eigen::ThreadPoolDevice, float>::operator()(int, int, int, int, int, int, int, int, int, int, int, int, int, float const*, float const*, float*, float*, bool, float*, float*)::{lambda(long, long)#1}::operator()(long, long) const (this=0x555555b0c560, start=7, limit=8) at /home/jyoung/git/npu/ONE/compute/cker/include/cker/eigen/depthwise_conv_op.h:592
* 19   Thread 0x7fffef16d700 (LWP 1326028) "test_cker" nnfw::cker::depthwise_conv_op::LaunchDepthwiseConvBackpropInputOp<Eigen::ThreadPoolDevice, float>::operator()(int, int, int, int, int, int, int, int, int, int, int, int, int, float const*, float const*, float*, float*, bool, float*, float*)::{lambda(long, long)#1}::operator()(long, long) const (this=0x555555b0c560, start=5, limit=6) at /home/jyoung/git/npu/ONE/compute/cker/include/cker/eigen/depthwise_conv_op.h:592
  20   Thread 0x7fffee96c700 (LWP 1326029) "test_cker" nnfw::cker::depthwise_conv_op::LaunchDepthwiseConvBackpropInputOp<Eigen::ThreadPoolDevice, float>::operator()(int, int, int, int, int, int, int, int, int, int, int, int, int, float const*, float const*, float*, float*, bool, float*, float*)::{lambda(long, long)#1}::operator()(long, long) const (this=0x555555b0c560, start=1, limit=2) at /home/jyoung/git/npu/ONE/compute/cker/include/cker/eigen/depthwise_conv_op.h:592
  21   Thread 0x7fffee16b700 (LWP 1326030) "test_cker" nnfw::cker::depthwise_conv_op::LaunchDepthwiseConvBackpropInputOp<Eigen::ThreadPoolDevice, float>::operator()(int, int, int, int, int, int, int, int, int, int, int, int, int, float const*, float const*, float*, float*, bool, float*, float*)::{lambda(long, long)#1}::operator()(long, long) const (this=0x555555b0c560, start=2, limit=3) at /home/jyoung/git/npu/ONE/compute/cker/include/cker/eigen/depthwise_conv_op.h:592
```

And in the multiple batch size case, the main thread is also used to execute the `f` function. So I added the `+1` for main thread.

Reply:
https://github.com/libigl/eigen/blob/1f05f51517ec4fd91eed711e0f89e97a7c028c0e/unsupported/Eigen/CXX11/src/Tensor/TensorDeviceThreadPool.h#L241-L242

You can find the code that directly executes the user function. Using or not using the main thread is implementation-dependent and not a set rule. Additionally, in the Eigen library, it uses the main thread, so I need to add an additional memory buffer for the main thread.

Reply:
1. our `ruy` backend provides a way to limit the max_threads from environment variables.

runtime/onert/backend/ruy/ExternalContext.h

```c
class ExternalContext
{
private:
  static const int kDefaultNumThreadpoolThreads = 4;

public:
  ExternalContext() : _ruy_context(new ::ruy::Context)
  {
    setMaxNumThreads(onert::util::getConfigInt(onert::util::config::RUY_THREADS));
  }

  void setMaxNumThreads(int max_num_threads)
  {
    const int target_num_threads =
      max_num_threads > -1 ? max_num_threads : kDefaultNumThreadpoolThreads;
    _ruy_context->set_max_num_threads(target_num_threads);
  }
```
If I remember correctly, the thread count does not include main threads. Main thread waits for worker threads.

2. I don't know much about eigen internal, but at a glance,

```c
    if (n <= 1 || numThreads() == 1 ||
        CostModel::numThreads(n, cost, static_cast<int>(numThreads())) == 1) {
      f(0, n);
      return;
    }
```

seems to be the case for 1 batch. I don't understand why it spawns 21 or more threads.

Reply:
> If I remember correctly, the thread count does not include main threads. Main thread waits for worker threads.

I didn't say that the thread count includes main thread. The thread count returns 20 (in x64 machine) and main thread is not included in that count. So I added `+1` for main thread. And it is an eigen library. This code has nothing to do with ruy backend behaviour.

> seems to be the case for 1 batch. I don't understand why it spawns 21 or more threads.

I already commented the code that executes the operation in main thread before the main thread waits.

Reply:
@jyoungyun 

> seems to be the case for 1 batch. I don't understand why it spawns 21 or more threads.

I agree with @glistening . I think the code including the code below does not explain why main thread should be added to the thread count.
```c++
    Barrier barrier(static_cast<unsigned int>(block_count));
    std::function<void(Index, Index)> handleRange;
    handleRange = [=, &handleRange, &barrier, &f](Index first, Index last) {
      if (last - first <= block_size) {
        // Single block or less, execute directly.
        f(first, last);
        barrier.Notify();
        return;
      }
      // Split into halves and submit to the pool.
      Index mid = first + divup((last - first) / 2, block_size) * block_size;
      pool_->Schedule([=, &handleRange]() { handleRange(mid, last); });
      handleRange(first, mid);
    };
    handleRange(0, n);
    barrier.Wait();
  }
```

Reply:
@ragmani In the above code, the main thread schedules for multi-threading and also executes the `handleRange` itself.

Reply:
> @glistening
> 
> > If I remember correctly, the thread count does not include main threads. Main thread waits for worker threads.
> 
> I didn't say that the thread count includes main thread. And that's why I added `1` to the thread counts. The thread count returns 20 (in x64 machine) and main thread is not included in that count. So I added `1` for main thread. And it is an eigen library. This code has nothing to do with ruy backend behaviour.

I guess I already understood what you say. I want to keep our code that uses external library consistent. A few years ago, there was some requirements to limit the total number of threads. I guess `onert::util::config::RUY_THREADS` make it easy to control how many threads will be allowed for ruy kernel. But this PR of eigen case is different. It is hard to control and understand how many threads are used/allowed for this kernel.

Reply:
@glistening 

> What is returned by `_dconv_kernel->getThreadCount()`? Is the number controllable?

It returns the core number. And we couldn't change this value.

Reply:
> @ragmani In the above code, the main thread schedules for multi-threading and also executes the handleRange itself.

You are right. Please ignore my comment.

Reply:
@jyoungyun 

> In the above code, the main thread schedules for multi-threading and also executes the `handleRange` itself.

It seems that you are right. Main thread may call the code below.

```c++
      if (last - first <= block_size) {
        // Single block or less, execute directly.
        f(first, last);
        barrier.Notify();
        return;
      }
```

Reply:
@glistening 
> I guess I already understood what you say. I want to keep our code that uses external library consistent. A few years ago, there was some requirement to limit the total number of threads. I guess onert::util::config::RUY_THREADS make it easy to control how many threads will be allowed for ruy kernel. Also, we can tune how many threads are better by changing the number of thread.

`cker` has already been using `Eigen::ThreadPoolDevice`, which has as many threads as the number of cores. For example, `Conv` kernel for inference, optimizer kernels for training, `BroadcastTo` kernel. If we only use kernels that easily controls the number of threads, we have to implement kernels ourselves to replace kernels that are hard to control. 

@jyoungyun 
Is there a way to apply the optimized `DepthwiseConv` kernels without calling `getThreadCount()`?

Reply:
If we controls the number of threads in `Eigen::ThreadPoolDevice`, we have to modify the code below:
https://github.com/Samsung/ONE/blob/2a4d8d60c77e45672ac3dfabf7970d6005074f93/compute/cker/include/cker/eigen/EigenSupport.h#L85-L114

Reply:
@jyoungyun and @ragmani Thank you for explanation. For Eigen, the number of thread is determined by the number of core on the device. It is not controllable like `ruy` backend.